pkg/kv/store_message.go

```diff
@@ -51,62 +51,133 @@ func (s *StoreMessage) DeleteMsg(ctx context.Context, path string) error {
+	return &MessageEntry{
+		Key:   "",
+		Value: msg.New().Interface(),
+	}
+}
+
+type MessageIterator interface {
+	Next() bool
+	Entry(key *string, value *protoreflect.ProtoMessage) error
+	Err() error
+	Close()
+}
+
+// PrimaryIterator MessageIterator implementation for primary key
+// The iterator iterates over the given prefix and returns the proto message and key
+type PrimaryIterator struct {
+	itr EntriesIterator
+	err error
+}
+
+func NewPrimaryIterator(ctx context.Context, store Store, prefix string) (*PrimaryIterator, error) {
 	itr, err := ScanPrefix(ctx, store, []byte(prefix))
 	if err != nil {
 		return nil, fmt.Errorf("create prefix iterator: %w", err)
 	}
-	return &MessageIterator{itr: itr, msgType: msgType}, nil
+	return &PrimaryIterator{itr: itr}, nil
 }
 
-func (m *MessageIterator) Next() bool {
+func (m *PrimaryIterator) Next() bool {
 	if m.err != nil {
 		return false
 	}
 	return m.itr.Next()
 }
 
-func (m *MessageIterator) Entry() *MessageEntry {
+func (m *PrimaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
 	entry := m.itr.Entry()
 	if entry == nil {
-		return nil
+		return ErrNotFound
 	}
-	msg := m.msgType.New().Interface()
-	err := proto.Unmarshal(entry.Value, msg)
-	if err != nil {
-		m.err = fmt.Errorf("unmarshal proto data for key %s: %w", string(entry.Key), err)
-		return nil
+	if value != nil {
+		err := proto.Unmarshal(entry.Value, *value)
+		if err != nil {
+			return fmt.Errorf("unmarshal proto data for key %s: %w", string(entry.Key), err)
+		}
 	}
-	return &MessageEntry{
-		Key:   string(entry.Key),
-		Value: msg,
+	if key != nil {
+		*key = string(entry.Key)
 	}
+	return nil
 }
 
-func (m *MessageIterator) Err() error {
+func (m *PrimaryIterator) Err() error {
 	if m.err != nil {
 		return m.err
 	}
 	return m.itr.Err()
 }
 
-func (m *MessageIterator) Close() {
+func (m *PrimaryIterator) Close() {
 	m.itr.Close()
 }
+
+// SecondaryIterator MessageIterator implementation for secondary key
+// The iterator iterates over the given prefix, extracts the primary key value from secondary key and then returns
+// the proto message and primary key
+type SecondaryIterator struct {
+	ctx   context.Context
+	itr   EntriesIterator
+	store Store
+	err   error
+}
+
+func NewSecondaryIterator(ctx context.Context, store Store, prefix string) (*SecondaryIterator, error) {
+	itr, err := ScanPrefix(ctx, store, []byte(prefix))
+	if err != nil {
+		return nil, fmt.Errorf("create prefix iterator: %w", err)
+	}
+	return &SecondaryIterator{ctx: ctx, itr: itr, store: store}, nil
+}
+
+func (s *SecondaryIterator) Next() bool {
+	if s.err != nil {
+		return false
+	}
+	return s.itr.Next()
+}
+
+func (s *SecondaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
+	secondary := s.itr.Entry()
+	if secondary == nil {
+		return ErrNotFound
+	}
+
+	if value != nil {
+		data, err := s.store.Get(s.ctx, secondary.Value)
+		if err != nil {
```

Comment:
Notice that ErrNotFound here needs to be ignored. Otherwise you may always find yourself in a failure loop.
Inconsistencies may occur in the KV world and the user (e.g. the action service) needs to be aware of that and handle gracefully to secure the operation of the service in all possible states.

Reply:
What do you think the expected behavior should be in this case? Should we iterate on secondary indices until a valid one is found? Should we return the ErrNotFound and let the caller handle this special case, should we return nil?